lib/lrama/grammar/rule_builder.rb

```diff
@@ -110,11 +110,16 @@ def process_rhs
           when Lrama::Lexer::Token::Ident
             @replaced_rhs << token
           when Lrama::Lexer::Token::Parameterizing
-            parameterizing = ParameterizingRules::Builder.new(token, @rule_counter, @lhs_tag, user_code, precedence_sym, line)
-            parameterizing.build.each do |r|
-              @parameterizing_rules << r
+            if parameterizing_resolver.defined?(token.s_value)
+              parameterizing_resolver.build_rules(token, @rule_counter, @lhs_tag, line)
+              @parameterizing_rules = @parameterizing_rules + parameterizing_resolver.rules
```

Comment:
`parameterizing_resolver.rules` stores all rules. Using `parameterizing_resolver.rules` in RuleBuilder causes multiple same rules.
`parameterizing_resolver.rules` is a cache for reusing same rules for same arguments, then `ParameterizingRuleResolver@rules` should be hidden from outside of class and `ParameterizingRuleResolver#build_rules` should return generated (or cached) rules corresponding to arguments.
For example

```
/*
 * This is comment for this file.
 */

%{
// Prologue
static int yylex(YYSTYPE *val, YYLTYPE *loc);
static int yyerror(YYLTYPE *loc, const char *str);

%}
%union {
    int i;
}

%token number

%rule defined_option(X): /* empty */
                       | X
                       ;

%rule defined_option2(X): /* empty */
                       | X
                       ;

%rule unused_define(X): /* empty */
                      | X
                      ;

%%

program         : defined_option(number) <i>
                | defined_option2(number) <i>
                ;

%%

static int yylex(YYSTYPE *yylval, YYLTYPE *loc) {
{
  return 0;
}

static int yyerror(YYLTYPE *loc, const char *str) {
{
  return 0;
}

int main(int argc, char *argv[])
{
}
```

```
$ bundle exec exe/lrama --trace=rules spec/fixtures/parameterizing_rules/user_defined.y

shift/reduce conflicts: 1 found
reduce/reduce conflicts: 1 found
Grammar rules:
$accept -> program, YYEOF
program -> defined_option_number
defined_option_number -> ε
defined_option_number -> number
program -> defined_option_number, defined_option2_number
defined_option_number -> ε
defined_option_number -> number
defined_option2_number -> ε
defined_option2_number -> number
```

Reply:
I believe the following modifications have been made together! Thanks🙏
https://github.com/ruby/lrama/pull/285#discussion_r1421707337